### PR TITLE
S1656: Add support for deconstruction

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/SelfAssignmentTest.cs [hunk 1]
```diff
@@ -26,20 +26,21 @@ namespace SonarAnalyzer.UnitTest.Rules
     [TestClass]
     public class SelfAssignmentTest
     {
+        private readonly VerifierBuilder builderCS= new VerifierBuilder<CS.SelfAssignment>();
+        private readonly VerifierBuilder builderVB= new VerifierBuilder<VB.SelfAssignment>();
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/SelfAssignmentTest.cs:30]: code style: one more space is needed before the equals sign.

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM - There are two spaces missing in the Test file.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SelfAssignment.cs [hunk 2]
```diff
@@ -32,25 +34,23 @@ public sealed class SelfAssignment : SelfAssignmentBase<SyntaxKind>
...
+                    var expression = (AssignmentExpressionSyntax)c.Node;
 
                     if (expression.Parent is InitializerExpressionSyntax)
                     {
                         return;
                     }
 
-                    if (CSharpEquivalenceChecker.AreEquivalent(expression.Left, expression.Right))
+                    var assigments = expression.MapAssignmentArguments().Where(x => CSharpEquivalenceChecker.AreEquivalent(x.Left, x.Right));
+                    foreach (var assigment in assigments)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SelfAssignment.cs:48]: I wouldn't give it a name, it's single use. I'd put the linq inside foreach

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**